test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,410 @@
+constexpr int kNumConnections = 10;
+constexpr int kNumExchangedMessages = 100;
+std::atomic<int> g_num_active_connections{0};
+
+// Returns a random message with bounded length.
+std::string GetNextSendMessage() {
+  static const char alphanum[] =
+      "0123456789"
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+      "abcdefghijklmnopqrstuvwxyz";
+  static std::random_device rd;
+  static std::seed_seq seed{rd()};
+  static std::mt19937 gen(seed);
+  static std::uniform_real_distribution<> dis(kMinMessageSize, kMaxMessageSize);
+  static grpc_core::Mutex g_mu;
+  std::string tmp_s;
+  int len;
+  {
+    grpc_core::MutexLock lock(&g_mu);
+    len = dis(gen);
+  }
+  tmp_s.reserve(len);
+  for (int i = 0; i < len; ++i) {
+    tmp_s += alphanum[rand() % (sizeof(alphanum) - 1)];
+  }
+  return tmp_s;
+}
+
+std::shared_ptr<EventEngine> GetPosixEE() {
+  static std::shared_ptr<EventEngine> posix_ee =
+      std::make_shared<PosixEventEngine>();
+  return posix_ee;
+}
+
+EventEngine* GetOracleEE() {
+  static EventEngine* oracle_ee = new PosixOracleEventEngine();
+  return oracle_ee;
+}
+
+class TestScheduler : public Scheduler {
+ public:
+  explicit TestScheduler(EventEngine* engine) : engine_(engine) {}
+  void Run(EventEngine::Closure* closure) override { engine_->Run(closure); }
+
+  void Run(absl::AnyInvocable<void()> cb) override {
+    engine_->Run(std::move(cb));
+  }
+
+ private:
+  EventEngine* engine_;
+};
+
+std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                     std::unique_ptr<EventEngine::Endpoint>>>
+CreateConnectedEndpoints(PosixEventPoller* poller, bool is_zero_copy_enabled,
+                         int num_connections) {
+  EXPECT_NE(GetOracleEE(), nullptr);
+  EXPECT_NE(GetPosixEE(), nullptr);
+  EXPECT_NE(poller, nullptr);
+  std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                       std::unique_ptr<EventEngine::Endpoint>>>
+      connections;
+  auto memory_quota = absl::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  EventEngine::ResolvedAddress resolved_addr =
+      URIToResolvedAddress(target_addr);
+  Promise<std::unique_ptr<EventEngine::Endpoint>> server_endpoint_promise;
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint_promise](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint_promise.Set(std::move(ep));
+      };
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  if (is_zero_copy_enabled) {
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_ENABLED, 1);
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_SEND_BYTES_THRESHOLD,
+                    kMinMessageSize);
+  }
+  ChannelArgsEndpointConfig config(args);
+  auto status = GetOracleEE()->CreateListener(
```

Comment:
Per a totw I linked you to earlier, I suggest:

```
auto listener = oracle_ee->CreateListener(...)
ASSERT_TRUE(listener.ok());
listener->Bind(...);
listener->Start(...);
```

Reply:
I couldn't use ASSERT_TRUE because the function returns a non void value. But I renamed the variable.